File: Android App/app/src/main/java/revert/shredsheets/models/TuningModel.java
package revert.shredsheets.models;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

import revert.shredsheets.enums.Keys;

public class TuningModel {
    public static Keys[] getStandardTuning() {
        return new Keys[]{Keys.E, Keys.B, Keys.G, Keys.D, Keys.A, Keys.E, Keys.B, Keys.F_Sharp, Keys.C_Sharp, Keys.G_Sharp, Keys.D_Sharp, Keys.A_Sharp};
    }

    public static Keys[] getDroppedDTuning() {
        return new Keys[]{Keys.E, Keys.B, Keys.G, Keys.D, Keys.A, Keys.D, Keys.A, Keys.D, Keys.A, Keys.D, Keys.A, Keys.D};
    }

    public static Keys[] getNewStandardTuning() {
        return new Keys[]{Keys.G, Keys.E, Keys.A, Keys.D, Keys.G, Keys.C, Keys.G, Keys.E, Keys.A, Keys.D, Keys.G, Keys.C};
    }

    public static Keys[] getOpenCTuning() {
        return new Keys[]{Keys.E, Keys.C, Keys.G, Keys.C, Keys.G, Keys.C, Keys.E, Keys.C, Keys.G, Keys.C, Keys.G, Keys.C};
    }

    public static Keys[] getOpenDTuning() {
        return new Keys[]{Keys.D, Keys.A, Keys.D, Keys.F_Sharp, Keys.A, Keys.D, Keys.D, Keys.A, Keys.D, Keys.F_Sharp, Keys.A, Keys.D};
    }

    public static Keys[] getOpenETuning() {
        return new Keys[]{Keys.E, Keys.B, Keys.E, Keys.G_Sharp, Keys.B, Keys.E, Keys.E, Keys.B, Keys.E, Keys.G_Sharp, Keys.B, Keys.E};
    }

    public static Keys[] getOpenFTuning() {
        return new Keys[]{Keys.C, Keys.F, Keys.C, Keys.F, Keys.A, Keys.F, Keys.C, Keys.F, Keys.C, Keys.F, Keys.A, Keys.F};
    }

    public static Keys[] getOpenGTuning() {
        return new Keys[]{Keys.D, Keys.G, Keys.D, Keys.G, Keys.B, Keys.D, Keys.D, Keys.G, Keys.D, Keys.G, Keys.B, Keys.D};
    }

    public static Keys[] getOpenATuning() {
        return new Keys[]{Keys.E, Keys.A, Keys.C_Sharp, Keys.E, Keys.A, Keys.E, Keys.E, Keys.A, Keys.C_Sharp, Keys.E, Keys.A, Keys.E};
    }

    public static Keys[] getOpenBTuning() {
        return new Keys[]{Keys.B, Keys.F_Sharp, Keys.B, Keys.F_Sharp, Keys.B, Keys.D_Sharp, Keys.B, Keys.F_Sharp, Keys.B, Keys.F_Sharp, Keys.B, Keys.D_Sharp};
    }

    public static Keys[] getUkuleleStandardTuning() {
        return new Keys[]{Keys.E, Keys.B, Keys.A, Keys.E, Keys.C, Keys.G, Keys.E, Keys.A, Keys.G, Keys.C, Keys.E, Keys.G};
    }

    public static Keys[] getFiveStringBassTuning() {
        return new Keys[]{Keys.B, Keys.G, Keys.D, Keys.A, Keys.E, Keys.B, Keys.G, Keys.D, Keys.A, Keys.E, Keys.B, Keys.G};
    }

    public static Keys[] getEightStringAlternateTuning() {
        return new Keys[]{Keys.E, Keys.B, Keys.G, Keys.D, Keys.A, Keys.E, Keys.B, Keys.E, Keys.D, Keys.A, Keys.E, Keys.B};
    }

    public static HashMap<Keys[], String> getTunings() {
        HashMap<Keys[], String> tunings = new HashMap<>(13);
        tunings.put(getStandardTuning(), "Standard Tuning");
        tunings.put(getDroppedDTuning(), "Drop D");
        tunings.put(getNewStandardTuning(), "New Standard");
        tunings.put(getOpenCTuning(), "Open C");
        tunings.put(getOpenDTuning(), "Open D");
        tunings.put(getOpenETuning(), "Open E");
        tunings.put(getOpenFTuning(), "Open F");
        tunings.put(getOpenGTuning(), "Open G");
        tunings.put(getOpenATuning(), "Open A");
        tunings.put(getOpenBTuning(), "Open B");
        tunings.put(getUkuleleStandardTuning(), "Ukulele Standard");
        tunings.put(getFiveStringBassTuning(), "5str Bass Standard");
        tunings.put(getEightStringAlternateTuning(), "8str Alternate");
        return tunings;
    }

    public static String getTuningName(Keys[] keys) {
        Set<Map.Entry<Keys[], String>> tunings = getTunings().entrySet();

        for (Map.Entry<Keys[], String> tuning : tunings) {
            if (Arrays.equals(keys, tuning.getKey())) {
                return tuning.getValue();
            }
        }

        StringBuilder tuningBuilder = new StringBuilder();
        for (Keys key : keys) {
            tuningBuilder.append(key);
        }
        return "Tuning: " + tuningBuilder.toString();
    }
}

File: Shredsheets 3/core/src/com/bm/shredsheets/models/TuningModel.kt
package com.bm.shredsheets.models

import com.bm.shredsheets.enums.MusicKeys
import com.bm.shredsheets.enums.setFrequency
import java.util.*

object TuningModel {
    val standardTuning: Array<MusicKeys>
        get() = arrayOf(MusicKeys.E.setFrequency(329.63f), MusicKeys.B.setFrequency(246.94f), MusicKeys.G.setFrequency(196.00f), MusicKeys.D.setFrequency(146.83f), MusicKeys.A.setFrequency(110.00f), MusicKeys.E.setFrequency(82.41f),
            MusicKeys.B.setFrequency(61.74f), MusicKeys.F_Sharp.setFrequency(23.12f))
    val droppedDTuning: Array<MusicKeys>
        get() = arrayOf(MusicKeys.E, MusicKeys.B, MusicKeys.G, MusicKeys.D, MusicKeys.A, MusicKeys.D, MusicKeys.A, MusicKeys.D)
    val newStandardTuning: Array<MusicKeys>
        get() = arrayOf(MusicKeys.G, MusicKeys.E, MusicKeys.A, MusicKeys.D, MusicKeys.G, MusicKeys.C, MusicKeys.G, MusicKeys.E)
    val openCTuning: Array<MusicKeys>
        get() = arrayOf(MusicKeys.E, MusicKeys.C, MusicKeys.G, MusicKeys.C, MusicKeys.G, MusicKeys.C, MusicKeys.E, MusicKeys.C)
    val openDTuning: Array<MusicKeys>
        get() = arrayOf(MusicKeys.D, MusicKeys.A, MusicKeys.D, MusicKeys.F_Sharp, MusicKeys.A, MusicKeys.D, MusicKeys.D, MusicKeys.A)
    val openETuning: Array<MusicKeys>
        get() = arrayOf(MusicKeys.E, MusicKeys.B, MusicKeys.E, MusicKeys.G_Sharp, MusicKeys.B, MusicKeys.E, MusicKeys.E, MusicKeys.B)
    val openFTuning: Array<MusicKeys>
        get() = arrayOf(MusicKeys.C, MusicKeys.F, MusicKeys.C, MusicKeys.F, MusicKeys.A, MusicKeys.F, MusicKeys.C, MusicKeys.F)
    val openGTuning: Array<MusicKeys>
        get() = arrayOf(MusicKeys.D, MusicKeys.G, MusicKeys.D, MusicKeys.G, MusicKeys.B, MusicKeys.D, MusicKeys.D, MusicKeys.G)
    val openATuning: Array<MusicKeys>
        get() = arrayOf(MusicKeys.E, MusicKeys.A, MusicKeys.C_Sharp, MusicKeys.E, MusicKeys.A, MusicKeys.E, MusicKeys.E, MusicKeys.A)
    val openBTuning: Array<MusicKeys>
        get() = arrayOf(MusicKeys.B, MusicKeys.F_Sharp, MusicKeys.B, MusicKeys.F_Sharp, MusicKeys.B, MusicKeys.D_Sharp, MusicKeys.B, MusicKeys.F_Sharp)
    val ukuleleStandardTuning: Array<MusicKeys>
        get() = arrayOf(MusicKeys.E, MusicKeys.B, MusicKeys.A, MusicKeys.E, MusicKeys.C, MusicKeys.G, MusicKeys.E, MusicKeys.A)
    val fiveStringBassTuning: Array<MusicKeys>
        get() = arrayOf(MusicKeys.B, MusicKeys.G, MusicKeys.D, MusicKeys.A, MusicKeys.E, MusicKeys.B, MusicKeys.G, MusicKeys.D)
    val eightStringAlternateTuning: Array<MusicKeys>
        get() = arrayOf(MusicKeys.E, MusicKeys.B, MusicKeys.G, MusicKeys.D, MusicKeys.A, MusicKeys.E, MusicKeys.B, MusicKeys.E)
    val tunings: HashMap<Array<MusicKeys>, String>
        get() {
            val tunings: HashMap<Array<MusicKeys>, String> = HashMap(13)
            tunings[standardTuning] = "Standard Tuning"
            tunings[droppedDTuning] = "Drop D"
            tunings[newStandardTuning] = "New Standard"
            tunings[openCTuning] = "Open C"
            tunings[openDTuning] = "Open D"
            tunings[openETuning] = "Open E"
            tunings[openFTuning] = "Open F"
            tunings[openGTuning] = "Open G"
            tunings[openATuning] = "Open A"
            tunings[openBTuning] = "Open B"
            tunings[ukuleleStandardTuning] = "Ukulele Standard"
            tunings[fiveStringBassTuning] = "5str Bass Standard"
            tunings[eightStringAlternateTuning] = "8str Alternate"
            return tunings
        }

    fun getTuningName(keys: Array<MusicKeys?>): String {
        val tunings: Set<Map.Entry<Array<MusicKeys>, String>> = tunings.entries
        for ((key, value) in tunings) {
            if (Arrays.equals(keys, key)) {
                return value
            }
        }
        val tuningBuilder = StringBuilder()
        for (key in keys) {
            tuningBuilder.append(key)
        }
        return "Tuning: $tuningBuilder"
    }
}
